---
title: " Intro to RNA-seq"
author: "Vinay Swamy"
date: "10/13/2020"
output:
  ioslides_presentation: 
    widescreen: true
---
<link rel="stylesheet" href="http://yandex.st/highlightjs/7.3/styles/default.min.css">
<script src="http://ajax.googleapis.com/ajax/libs/jquery/1.9.0/jquery.min.js"></script>
<script src="http://yandex.st/highlightjs/7.3/highlight.min.js"></script>
<script>
$(document).ready(function() {
  $('pre code').each(function(i, e) {hljs.highlightBlock(e)});
});
</script>
<style>
slides > slide {
  overflow-y: auto !important;
}
.container{
    display: flex;
}
.col{
    flex: 1;
}
</style>

```{r setup, include = F}
library(knitr)
library(tidyverse)
#opts_chunk$set(echo = F, fig.height = 5, fig.width = 5)
opts_knit$set(root.dir = '~/personal/BIOF045-Intro-to-RNA-seq/')
```


## Schedule
- Overview on RNA-seq
- How to perform transcript quantification with Salmon
- Using R for making Plots
- Exploring RNA-seq data 
- Differential Gene expression

## RNA Biology -  The Central Dogma
<div class="container">
<div class="col">
```{r}
include_graphics('src/dna_central_dogma_yourgenome.png')
```

</div>

<div class="col">
- In transcription, distinct parts of the genome are transcribed into immature RNA. 
- `inline exampl`
</div>
</div> 

## RNA processing 

<div class="container">
<div class="col">
```{r}
include_graphics('src/rna_isoforms.gif')
```

</div>

<div class="col">
####  
- RNA processing mechanisms convert immature RNA to mature RNA (mRNA)
- RNA processing mechanisms include polyadenylation and splicing
- Exons - regions of immature RNA that are retained in mRNA
- introns - regions of immature RNA that are **NOT** retained in mRNA
</div>
</div>


## One gene, multiple isoforms
<div class="container">
<div class="col">
```{r}
include_graphics('src/alt_splice.jpg')
```
</div>
<div class="col">
- the locus associated with a single gene often leads to multiple distinct isoforms
- This occurs because each RNA processing mechanism can lead to the formation of different gene isoforms
- For example, alternative splicing leads to the inclusion/exclusion of different exons, and can lead encoded proteins with distinct functions 
- mRNA transcripts are the distinct units of transcription, not genes
</div>
</div> 

## High throughput RNA sequencing - RNA-seq

<div class="container">
<div class="col">
```{r}
include_graphics('src/rnaseq_lib_prep.png')
```
</div>
<div class="col">
- In RNA-seq, mRNA transcripts are converted to cDNA
- cDNA is fragmented. Steps up to here are commonly referred to as library preparation
- Fragments are amplified with PCR and then sequenced. Each PCR cycle there is a chance of introducing errors
- Amplified fragments are called "Reads". Reads are typically 50-150bp long. Generally there's around 50 Million reads per sample
- reads are then computationally assigned to the original mRNA transcript they come from 
</div>
</div> 

## Library Selection Method

<div class="container">
<div class="col">
```{r}
include_graphics('src/types_of_rna.png')
```
</div>
<div class="col">
- There are multiple types of RNA
- RNA-seq generally refers to mRNA only, so part of the library prep procedure is to remove unwanted types
- polyA selection 
  - pros: cheap, and efficient. Most commonly used method
  - cons: removes ncRNA and introduces a bias for the 3' end of transcripts
- riboRNA(rRNA) depletion
  - pros: only removes rRNA and tRNA, no coverage bias, useful for prokaryotic organisms
  - cons: expensive
</div>
</div> 

## The Spliced alignment problem

<div class="container">
<div class="col">
```{r}
include_graphics('src/spliced_aln.png')
```
</div>
<div class="col">
- The challenge with aligning RNA-seq reads arises from the fact that reads spanning exon-exon junctions can't be directly mapped to the genome
- aligners for RNA-seq most account for this problem, often at the expense of increased computing resources
- typical spliced aligners: TopHat, STAR, Hisat2
- these tools align reads to the genome
</div>
</div> 

## Library Type
- stranded  OR
- unstranded
          AND
- single OR
- paired 

## Transcript Quantification
- With transcript quantification, reads are assigned to individual mRNA transcripts, not the genome. This eliminates the spliced alignment problem 
- In transcript quantification, reads are not precisely aligned, but rather are assigned to where they most likely should be aligned (pseudoalignment)
- Early iterations of transcript aligners were not as accurate as conventional genome based alignment, but newer versions have been show to as accurate as genome based alignment
- Transcript quantification requires **significantly** fewer resources than spliced alignment. 
- typical tools: Salmon, Kallisto. We'll be using salmon for this course 

## Using Salmon
- While the code I show is specifically for Salmon, the general process is similar for any transcriptome quantifier
- What you'll need to run Salmon
  - FASTA(.fa/.fasta) file for all RNA transcripts for your organism, ie [human](https://www.gencodegenes.org/human/). I'll refer to this as `transcripts.fa`
  - FASTA(.fa/.fasta) file for genome sequence for your organism I'll refer to this as `transcripts.fa`
  - FASTQ(.fq/.fastq) file(s) for your RNA-seq data. Me example is with paired end data, which I'll refer to as `sample_left.fq` and `sample_right.fq`

## Preparing an Index for Salmon
- the index is a computationally efficient representation of the distinct RNA transcripts in `transcripts.fa`
- the index preparation for Salmon is slightly different, and requires the preparation of decoy sequences. decoy sequences improve the accuracy of transcriptome quantification
```{bash, eval= F, echo = T}
grep "^>" genome.fa | cut -d " " -f 1 | sed  -e 's/>//g'  > decoys.txt # get names of chromosomes 
cat transcripts.fa genome.fa > transcript_genome_merged.fa # merge two fastas together
salmon index -t gentrome.fa.gz -d decoys.txt -p 12 -i salmon_human_index --gencode # use salmon to build the index
```

## Quantifying Transcripts 
- We next use our index and fastq files(s) for quantification
- This is a typical command
```{bash, eval = F, echo = T}
salmon quant  \
  --threads 8 \ # number of process to use
  --libType A \  # stranded or unstranded library. Use "A" to allow salmon to infer
  -1 sample_left.fq \ # fastqs; use "-r sample.fq" for single end data 
  -2 sample_right.fq \
  --index salmon_human_index \ # name of th index we built earlier
  --seqBias \ # Perform sequence-specific bias correction.
  --gcBias \  # Perform fragment GC bias correction.
  --posBias \ # Perform positional bias correction.
  --output sample_quantified/ # output folder. the actual file with the quantification will be named "quant.sf"
```

## Output
- the `quant.sf` file produced looks something like this 
```{r, echo = F}
txt <- 'Name    Length  EffectiveLength TPM     NumReads
DNTX_00000017   12501   9261.044        2.532237        2529.632
DNTX_00000019   3218    1783.769        12.960541       2493.762
DNTX_00000020   39670   27993.255       1.370513        4138.375
DNTX_00000018   26868   17512.212       2.156619        4073.875
DNTX_00000034   12032   7527.305        0.713575        579.392
DNTX_00000033   8417    6031.772        0.725138        471.801
DNTX_00000035   3349    1579.033        3.459229        589.201'
read.table(text = txt, header = T ) %>% kable()
```


## Analysis of RNA-seq data in R
- R is the most common language for (bulk)RNA-seq analysis by a wide margin
- Most of the packages for RNA-seq analysis are hosted on `Bioconductor`, which is a repository similar to CRAN, but more biologically focused. 
- we can get access to `Bioconductor` by installing the `BiocManager` library
```{r eval=F}
install.packages('BiocManager')
```
- we can then use `BiocManager::install` to download packages from `Bioconductor`

## Analysis of RNA-seq data in R
- there are a *lot* of tools for analyzing RNA-seq data. 
- We'll be covering `DESeq2` for this course, because it is by far the simplest to use
- tools like `edgeR` and `limma` are a little more powerful, but are harder to use.
- download `DESeq2` with `BiocManager::install('DESeq2')`


## Normalization of RNA-seq data
- Normalization of RNA-seq data is a *critical* step for its analysis
- Aligned RNA-seq data is presented as counts, where each read is a single count 
- The two most important parts of RNA-seq normalization  is "between-sample normalization" and "between transcript normalization"

## Between Sample Normalization 
- Library size refers to the number of reads in a given sample
- Library size is generally not indicative of the total amount of gene expression within the sample, but is influenced heavily by the PCR amplification step 
- And so when we compare samples we want to ensure that the number of reads is uniform across samples 

## Between transcript normalization
- mRNA transcripts have a wide range of lengths, and the number of counts per transcript is *directly* correlated with transcript length
- Consider two transcripts, A and B. 
   - A and B are expressed at the exact same levels .A is 1000 bp long, and B is 10000bp long. 
   - We divide each transcript into uniform 100bp fragments, producing 10 fragments for A(1000/100 = 10) and 100 fragments for B (10000/100 = 100)
   - We then each amplify each fragment 100 times. For A this gives 10 x 100 = 1000 reads, and for B this gives 100 *100 = 10000 reads
   - And so despite being expressed at the same level, transcript B has more counts and appears to be expressed higher. 
   - to combat this, we normalize the expression of each transcript by its length. For A this is 1000 reads/ 1000 bp = 1, and B 10000 reads/ 10000bp = 1

## Transcripts per Million reads (TPM)
- Combines both normalization methods, accounting for both transcript length and library size 
- Formally Let TPM is $TPM_t= \frac{c_t/l_t}S * 1000000$ where
  - $S=\sum_{t\ \in\ T} c_t/l_t$
  - $c_t$ = total number of counts(reads) for a  transcript *t*, 
  - $l_t$ = length of *t* in kilobases
  - $T$ = all transcripts quantified in the entire sample
- TPM values are provided as part of Salmon's output
  
## RNA-seq analyses methods use more robust methods for normalization
- RNA-seq analysis tools like `edgeR` and `DESeq2` use a more sophisticated approach for normalization, but generally follow the principles mentioned earlier
- Because of this, **ALWAYS** pass *raw* counts to these tools, and allow them to normalize for you

## Metadata
- For RNA-seq tool, you will need metadata about your samples; At the very least the metadata should include the sample names and experimental conditions.
```{r, message=F}
sample_table <- read_csv('src/sample_table.csv') 
kable(sample_table)
```

## Metadata

- it can also be useful to add paths to each file in the metadata file, so you can remember where they are on your computer


## Reading RNA-seq data into R
- the `tximport` package provides a uniform interface to read RNA-seq data into R. 
- It also can aggregate transcript level expression to gene level expression. 
- install the required packages this with `BiocManger::install(c('tximport', 'rtracklayer'))`

## Using `tximport`
```{r eval = F, echo = T}
library(tximport)
tximport(
    files, # a vector containing file paths to your "quant.sf" files
    type, # the type of tools used to generate the data ie "salmon" or "kallisto"
    txOut,# whether or not to return transcript level expression
    tx2gene,# a dataframe that maps transcript_id (first column) to gene_ids (second column)
    ignoreTxVersion# many transcript id's are versioned based on release. 
                   # set this to true to ignore versions
)
```

## Creating a `tx2gene`
- many tutorials recommend getting annotations through R-based tools, like `BiomaRt`
- I've found that its much easier to simply download the appropriate transcript annotation file(.GTF) from the same source you obtained your transcript fasta from
- for example with,  [gencode](https://www.gencodegenes.org/human/)
- gtfs can be read directly in R to with the `rtracklayer` package
```{recho = T, eval = F}
library(dplyr)
gtf <- rtracklayer::readGFF('path/to/gtf')
tx2gene <- gtf %>% 
  filter(type == 'transcript') %>% 
  select(transcript_id, gene_id) %>% 
  distinct()
```

## Gene level vs Transcript level expression
- It's much easier to work with gene level data than transcript level data
- We'll cover working with gene level data first, and if we have time discuss working with transcript level data 

## Loading data into `DESeq2`
- DESeq2 allows you to load data directly using the output of tximport
```{r, eval = F}
library(DESeq2)
DESeqDataSetFromTximport(  
  txi = # the  output from 
  colData = # a data fram containing metadata about each sample 
  design = # the experimental design of the study. Must be a column in metadata
  )
```

```{r include = F}
library("tximport")
library("readr")
library("tximportData")
library("DESeq2")
dir <- system.file("extdata", package="tximportData")
samples <- read.table(file.path(dir,"samples.txt"), header=TRUE)
samples$condition <- factor(rep(c("A","B"),each=3))
rownames(samples) <- samples$run
samples[,c("pop","center","run","condition")]
files <- file.path(dir,"salmon", samples$run, "quant.sf.gz")
names(files) <- samples$run
tx2gene <- read_csv(file.path(dir, "tx2gene.gencode.v27.csv"))
txi <- tximport(files, type="salmon", tx2gene=tx2gene)
```


## Study Designs
- A study design is based on the experimental conditions you are testing. 
- In R, this is a vector of conditions that corresponds to experimental condition of each sample. 
```{r echo = T}
samples$condition
```
- the design vector must be a column in the `colData` dataframe, and we refer to designs using the `~` operator,
- the design is used to signify that the variation in the observed in the data can best described by the design vector 

```{r echo=T}
ddsTxi <- DESeqDataSetFromTximport(txi,
                                   colData = samples,
                                   design = ~condition)
```

## Working with `DESeqDataSet` objects 
- `DESeq` uses a custom type of R object called a `DESeqDataSet`
```{r}
class(ddsTxi)
```
- this a fancy container to store multiple levels of data using the `$` operator on a `DESeqDataSet` object allows you to access columns in the dataframe you passed to colData
```{r}
ddsTxi$condition
```
## Working with `DESeqDataSet` objects 
- raw counts can be accessed with the `assay` function
```{r}
head(assay(ddsTxi))
```

## Analysis with `DESeq2`
- The general workflow of an analysis in RNA-seq experiment is the following
  - prefilter data
  - normalize for library size
  - account for technical variation
  - transforms count data
  - explore data
  - test for differential expression

## prefiltering data
- this step removes genes that have few counts. This is a always a good step, as it improves the statistical power of your tests, 
```{r}
dim(ddsTxi)
```

```{r}
minimally_expressed <- rowSums(assay(ddsTxi)) >=25
ddsTxi_filtered <- ddsTxi[minimally_expressed,]
dim(ddsTxi_filtered)
```

## library size calculations
```{r echo=F}
dds <- estimateSizeFactors(ddsTxi_filtered)
```
- `DESeq` first calculates the library size.  Remember that this is analogous to the total number of per sample
- the count data is then scaled based on library size

## Estimating Dispersions
```{r, echo = F}
dds <- estimateDispersions(dds)
```
- This is basically accounting for technical variability between expression of individual genes(its actually lot more complicated than that but beyond the scope of this course)
- for a more through explanation, see the documentation for the `estimateDispersions`
function


## Transforming Count Data
- raw RNA-seq counts are good for statistical models, but bad for visualization. These counts are typically transformed to reduce the large range between count values for different genes
- This is typically done by `log2` transforming the counts, ie `log2(counts + 1)`. 
- The 1 is added so counts less than stay above 0

## Transforming Count Data with DESeq2
- `DESeq` provides a slightly fancier way to log transform called `rlog`
```{r}
rlds <- rlog(dds)
```
- note that this returns a separate object than what we used earlier/ will use for differential expression


## Exploring data
- At this stage, we can begin exploring data.
- This is generally a QC step, to make sure that the data looks clean.

## A quick primer on dimensions 
- Generally, when people refer the "dimension" of data, they refer to the number of measurement, or variables,  within the data 
- We'll be using qPCR as an example. Lets say I measure the expression of 1 gene across 2 conditions, 
```{r include = F}
exp <- tibble(Condition = c(rep('wt', 3), rep('ko', 3)),
       Replicate = c(rep(1:3, 2)), 
       Gene_A = rnorm(6, 1))
exp2 <-  tibble(Condition = c(rep('wt', 3), rep('ko', 3)),
       Replicate = c(rep(1:3, 2)), 
       Gene_A = rnorm(6, 1), 
       Gene_B = rnorm(6, 1))
```
```{r}
exp
```
- this data has 6 observations, with 1 measurement, Gene_A, and so has one dimension

## A quick primer on dimensions 
```{r}
exp2
```
- Same number of observations, but with 2 measurement, and so this data has two dimension
- And so in a typical RNA-seq experiment where you are measuring 20K genes, you would have 20K dimensions
- However, note that RNA-seq is stored as variables x observations ( gene x sample)

## Exploring data with Principal Component Analysis
- Principal Component Analysis(PCA) is a method for reducing the dimensions from the data
- PCA computes a reduced dimensional space that maximally preserves the variation present in the original data. 
- We can use this reduced dimensional space to speed up computations, or for visualizations 

## PCA in DESeq2
- generating PCA plots is built into DESeq2. Simply pass a transformed data object to `plotPCA`
```{r}
plotPCA(rlds)
```

## Caution in interpreting PCA plots 
- Always pay attention to the amount of variance in each principal component
```{r}
library(ggplot2)
df <- tibble(x=c(-7, 3, -1, 10), 
       y=c(-1,-2, 2, 1 ), 
       condition = c('a', 'a', 'b', 'b'))
ggplot(df) + 
  geom_point(aes(x=y, y=x, color = condition)) + 
  xlab('PC1 45% variance explained') + 
  ylab('PC2 30% variance explained')

```


## Correlation Heatmaps
- Correlation heatmaps are a more quantitative way to examine similarity between samples 
- in a correlation heatmap, a correlation is calculated pairwise between samples
- correlations are a measure of similarity between samples. 1 is a perfect positive correlation, and -1 is a perfect negative correlation
```{r, nead}
suppressMessages(library(pheatmap))
cor_mat <- rlds %>% assay %>% cor( method = 'spearman')
# this is the same as cor(assay(rld))
pheatmap(cor_mat, 
         display_numbers = T,
         number_format='%.4f', 
         annotation_row = samples %>% select(condition))
```

## Correlation Heatmaps
- Because you get a quantitative measure of how similar samples are to each other, this is generally better for QC than PCA

## Testing For Differential Expression
- Differential expression is a statistical method for calculating whether the change in a gene's expression between two conditions is statistically significant 
-  there are a couple key ideas to consider

## Multiple Testing Correction
- Consider flipping a coin 10000 times, with a 50/50 probability. 
- you would expect the number of heads to be close to 5000, ie  a there's a 50% chance having half the population of flips landing head. 
- what would the probability of getting *at least* 1000 heads out of  -(its basically 1)
- This is because each coin flip is not an independent 
- When we perform a test for differential expression, we run a single statistical test for each gene. 
- Statistical tests are a measure of probability IE "the probability that we observe this change by chance"
- And so because we are running so many tests, we have a greater chance of finding something that might be statistically significant
- Part of this come from the fact that RNA-seq is extremely noisy, and so we want to be absolutely sure that what we find is statistically significant

## The actual statistical test
- The actual testing is done by fitting a linear model to the data, and then applying a statistical test to this model.
- There are multiple statistical tests you can use once this model is fit, the one I prefer is the Likelihood ratio test
- The likelihood model compares the a model with the design you specified (`~condition`) aka the full model,
to a second model, aka the reduced. At its simplest level this model is just ~1, which means that the best way to describe the data is no model, ie there is just random variation in the data. 

## The actual statistical test

```{r}
dds <-nbinomLRT(dds, full = design(dds), reduced = ~1)
```
- we can access the results using the `results` function
```{r}
diff_exp_results <-  results(dds) %>% as.data.frame
head(diff_exp_results)
```


## How to pick differentially expressed genes

- the two major criteria for considering a gene as differentially expressed are the adjusted p-value(padj) and fold change 
- fold change is the expression  of GeneA_conditionX / GeneA_conditionY. 
- We take the log of this value, so cases where GeneA_conditionX < GeneA_conditionY become negative and are on a similar scale as positive fold change
- A gene is differentially expressed if its pvalue is less than a specified value, and the absolute value of the log2fc is greater than a specified value

## selecting differentially expressed genes

```{r}
pvalue_cutoff = .05
l2fc_cutoff = 2
diff_exp_filtered <-  diff_exp_results %>% 
  rownames_to_column('gene_id') %>% 
  filter(padj < pvalue_cutoff, 
         abs(log2FoldChange) > l2fc_cutoff)
head(diff_exp_filtered)
```
- you can choose basically any value for `pvalue_cutoff` and `l2fc`

## some caveats for RNA-seq
- RNA-seq is by no means a definitive measure of gene expression. While it can be highly accurate, its equally as error prone
- RNA expression levels do not always line up with protein
- Always let biology inform your decision making when working with RNA-seq. 
- For example, if you have samples for different experimentally different that are biologically distinct IE morphologically, molecularly, functionally, etc, but the RNA-seq data doesn't exactly match that up, run an analysis again with more relaxed filtering values.


## Downstream of DGE
- We can use these differential expression results for gene set enrichment or for visualization
- Gene set enrichment determines how many of your differentially expressed genes are in a list of genes that correspond to a specific biological process or function.
- These processes are some times referred to as ontologies Ontologies, and exist in biological data bases 
- In order to do this, we need to map our gene ids to gene names 


## Gene IDs and gene names
- There is not a 1:1 mapping between ID's and names; multiple IDs refer to the same gene name, and so its generally better to use gene_ids over gene names for tasks like quantification and differential gene expression
- the best way to get gene names from annotations is by using an annotation file, like 



